src/core/src/package_managers/TdnfPackageManager.py

```diff
@@ -0,0 +1,458 @@
+
+    def extract_packages_and_versions_including_duplicates(self, output):
+        """Returns packages and versions from given output"""
+        # todo: verify usage
+        self.composite_logger.log_verbose("[TDNF] Extracting package and version data...")
+        packages = []
+        versions = []
+        package_extensions = Constants.SUPPORTED_PACKAGE_ARCH
+
+        def is_package(chunk):
+            # Using a list comprehension to determine if chunk is a package
+            return len([p for p in package_extensions if p in chunk]) == 1
+
+        lines = output.strip().split('\n')
+
+        for line_index in range(0, len(lines)):
+            # Do not install Obsoleting Packages. The obsoleting packages list comes towards end in the output.
+            if lines[line_index].strip().startswith("Obsoleting Packages"):
+                break
+
+            line = re.split(r'\s+', lines[line_index].strip())
+            next_line = []
+
+            if line_index < len(lines) - 1:
+                next_line = re.split(r'\s+', lines[line_index + 1].strip())
+
+            # If we run into a length of 3, we'll accept it and continue
+            if len(line) == 3 and is_package(line[0]):
+                packages.append(self.get_product_name(line[0]))
+                versions.append(line[1])
+            # We will handle these two edge cases where the output is on
+            # two different lines and treat them as one line
+            elif len(line) == 1 and len(next_line) == 2 and is_package(line[0]):
+                packages.append(self.get_product_name(line[0]))
+                versions.append(next_line[0])
+                line_index += 1
+            elif len(line) == 2 and len(next_line) == 1 and is_package(line[0]):
+                packages.append(self.get_product_name(line[0]))
+                versions.append(line[1])
+                line_index += 1
+            else:
+                self.composite_logger.log_verbose("[TDNF] > Inapplicable line (" + str(line_index) + "): " + lines[line_index])
+
+        return packages, versions
+    # endregion
+    # endregion
+
+    # region Install Updates
+    def get_composite_package_identifier(self, package, package_version):
+        # todo: need to change this
+        package_without_arch, arch = self.get_product_name_and_arch(package)
+        package_identifier = package_without_arch + '-' + self.get_package_version_without_epoch(package_version)
+        if arch is not None:
+            package_identifier += arch
+        return package_identifier
+
+    def install_updates_fail_safe(self, excluded_packages):
+        excluded_string = ""
+        for excluded_package in excluded_packages:
+            excluded_string += excluded_package + ' '
+        cmd = self.all_but_excluded_upgrade_cmd + excluded_string
+
+        self.composite_logger.log_debug("[TDNF][FAIL SAFE MODE] UPDATING PACKAGES USING COMMAND: " + cmd)
+        self.invoke_package_manager(cmd)
```

Comment:
These are all critical code paths in a full package manager implementation. A test matrix must include exercising them for correctness and applicability. Now all implementations are equal (or needed equally) between package managers and this seems like a like-for-like copy, hence the flagging.

Reply:
Good catch, removed this one. Have added all test logs in this PR description, will add a test matrix for easier readability